TextFieldEffects/TextFieldEffects/HoshiTextField.swift

```diff
@@ -12,178 +12,226 @@ import UIKit
+    }
+  }
+  
+  override open var placeholder: String? {
+    didSet {
+      updatePlaceholder()
+    }
+  }
+
+  override open var bounds: CGRect {
+    didSet {
+      frame = CGRect(x: frame.origin.x,
+                     y: frame.origin.y,
+                     width: bounds.size.width,
+                     height: height)
+    }
+  }
+
+  override open func awakeFromNib() {
+    super.awakeFromNib()
+    initPlaceholderFont()
+  }
+
+  private let borderThickness: (active: CGFloat, inactive: CGFloat) = (active: 2, inactive: 0.5)
+  private let placeholderInsets = CGPoint(x: 0, y: 25)
+  private let textFieldInsets = CGPoint(x: 0, y: 6)
+  private let inactiveBorderLayer = CALayer()
+  private let activeBorderLayer = CALayer()
+  private var activePlaceholderPoint: CGPoint = CGPoint(x: 0, y: -2)
+  private var inactivePlaceholderPoint: CGPoint {
+    get {
+      if let font = font, font.pointSize > 15 {
+        return CGPoint(x: 0, y: 14)
+      } else {
+        return CGPoint(x: 0, y: 20)
+      }
+    }
+  }
 
+  private var placeholderLabelOriginalText: String?
+  private var placeholderFont: UIFont?
+  private var height: CGFloat = 45
+
+  open func showError(message: String) {
+    placeholderLabelOriginalText = placeholderLabel.text
+    placeholderLabel.textColor = borderActiveColor
+    placeholderLabel.text = message
+    activeBorderLayer.frame = self.rectForBorder(self.borderThickness.active)
+    activeBorderLayer.isHidden = false
+    placeholderLabel.sizeToFit()
+  }
+
+  open func hideError() {
+    placeholderLabel.text = placeholderLabelOriginalText != nil ? placeholderLabelOriginalText : placeholderLabel.text
+    placeholderLabel.textColor = placeholderColor
+    activeBorderLayer.isHidden = true
+    placeholderLabel.sizeToFit()
+  }
+
+  // MARK: - TextFieldEffects
+
+  override open func drawViewsForRect(_ rect: CGRect) {
+    frame = CGRect(x: frame.origin.x,
+                   y: frame.origin.y,
+                   width: rect.size.width ,
+                   height: height)
+    configurePlaceholderLabelFrame()
+    configurePlaceholderFont()
+    updateBorder()
+    updatePlaceholder()
+
+    layer.addSublayer(inactiveBorderLayer)
+    layer.addSublayer(activeBorderLayer)
+    addSubview(placeholderLabel)
+  }
+
+  override open func animateViewsForTextEntry() {
+    if text!.isEmpty {
+      UIView.animate(withDuration: 0.35,
+                     delay: 0.0,
+                     usingSpringWithDamping: 0.8,
+                     initialSpringVelocity: 1.0,
+                     options: .beginFromCurrentState,
+                     animations: ({
+                      self.placeholderLabel.transform = CGAffineTransform(scaleX: 0.75, y: 0.75)
+                      self.placeholderLabel.font = self.placeholderFont
+                      self.placeholderLabel.sizeToFit()
+                      self.placeholderLabel.frame.origin = self.activePlaceholderPoint
+                     }), completion: { _ in
```

Comment:
I think you can use`[weak self] _  in` in completion block so there won't be strong reference cycle here.

Reply:
I need some more details to understand that strong reference cycle, I fail to see it.

HoshiTextField doesn't have a strong reference to the animation, so no retain cycle risk as far as I know

https://stackoverflow.com/a/27019834